hazelcast/src/main/java/com/hazelcast/internal/config/override/ExternalConfigurationOverride.java

```diff
@@ -0,0 +1,85 @@
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hazelcast.internal.config.override;
+
+import com.hazelcast.client.config.ClientConfig;
+import com.hazelcast.client.config.impl.YamlClientDomConfigProcessor;
+import com.hazelcast.config.Config;
+import com.hazelcast.internal.config.YamlMemberDomConfigProcessor;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.logging.Logger;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.function.BiConsumer;
+
+import static com.hazelcast.internal.config.override.PropertiesToNodeConverter.propsToNode;
+import static java.lang.String.format;
+
+/**
+ * A class used to process external configuration overrides coming from environment variables/system properties,
+ * and applying them over existing {@link Config}/{@link ClientConfig} configuration
+ */
+public class ExternalConfigurationOverride {
+
+    private static final ILogger LOGGER = Logger.getLogger(ExternalConfigurationOverride.class);
+
+    public Config overwriteMemberConfig(Config config) {
+        return overwrite(config, (provider, c) -> {
+              try {
+                  new YamlMemberDomConfigProcessor(true, c)
+                    .buildConfig(new ConfigOverrideElementAdapter(propsToNode(provider.properties())));
+              } catch (Exception e) {
+                  throw new RuntimeException(e);
+              }
+          },
+          new EnvConfigProvider(EnvVariablesConfigParser.member()),
+          new SystemPropertiesConfigProvider(SystemPropertiesConfigParser.member()));
+    }
+
+    public ClientConfig overwriteClientConfig(ClientConfig config) {
+        return overwrite(config, (provider, c) -> {
+              try {
+                  new YamlClientDomConfigProcessor(true, c)
+                    .buildConfig(new ConfigOverrideElementAdapter(propsToNode(provider.properties())));
+              } catch (Exception e) {
+                  throw new RuntimeException(e);
+              }
+          },
+          new EnvConfigProvider(EnvVariablesConfigParser.client()),
+          new SystemPropertiesConfigProvider(SystemPropertiesConfigParser.client()));
+    }
+
+    private <T> T overwrite(T config, BiConsumer<ConfigProvider, T> configProcessor, ConfigProvider... providers) {
+        ConfigOverrideValidator.validate(new HashSet<>(Arrays.asList(providers)));
+
+        for (ConfigProvider configProvider : providers) {
+            Map<String, String> properties = configProvider.properties();
+            properties.forEach((k, v) -> {
+                LOGGER.info(format("Detected external configuration overrides in %s: [%s]",
```

Comment:
Can we print one log statement instead with all of the overrides instead?

Reply:
I assume we want to minimize the number of lines? I came up with something like this:

> Detected external configuration overrides in environment variables: [hazelcast.network.join.autodetection.enabled=false,hazelcast.executorservice.custom.poolsize=42,hazelcast.clustername=test,hazelcast.cache.default.keytype.classname=java.lang.Object]

Reply:
Yes, that, and the PR description shows this as an example:
```
INFO: Detected external configuration overrides in environment variables: 
hazelcast.network.join.multicast.enabled=false
hazelcast.network.join.tcpip.enabled=false
hazelcast.clustername=foo
```